Add blog post on declarative plugins (#3636)
openbao/openbao: website/content/blog/2026-07-30-declarative-plugins.md

---

> title: "OpenBao Features - Declarative Plugin Distribution"

Reviewer: It's more of a post on OCI-based plugin distribution. I don't think distribution is in any way declarative. The configuration is declarative, but also covers a broader scope than OCI.

Author: I dropped `Distribution` from here.

| Type            | Description                                                                                                                                                                   |
| :-------------- | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Auth**        | Allow bringing sources of identity--JWTs, certificates, passwords--and exchanging them for OpenBao tokens.                                                                    |
| **Secret**      | Allow generating credentials (whether passwords for databases, PKI certificates, static secrets, and more) and protected through OpenBao's tokens and authorization policies. |
| **Database**    | Implement database-specific logic for updating and managing access for use with the `database` dynamic secret engine.                                                         |
| **Auto-Unseal** | Implement a specialized interfaces for using HSM and KMS devices (such as YubiHSM or GCP CloudKMS) for auto-unseal.                                                           |

> | **KMS**         | Implement a generic interface for using HSm and KMS devices within OpenBao's secret engines. _Coming in OpenBao v2.7.x_.                                                      |

Reviewer 2: ```suggestion
| **KMS**         | Implement a generic interface for using HSM and KMS devices within OpenBao's secret engines. _Coming in OpenBao v2.7.x_.                                                      |
```

> | **KMS**         | Implement a generic interface for using HSm and KMS devices within OpenBao's secret engines. _Coming in OpenBao v2.7.x_.                                                      |

Reviewer: I mean, yes, but they ship as one package. Perhaps defer external key additions to a later statement, outside of this table, since it is announced as "current"? Auto-Unseal plugins are already known as KMS plugins.

Author: Revised, combining the two together.

Plugins have two primary classifications:

| Classification | Description                                                                                                                                                |
| :------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | **External**   | Distributed as stand-alone binaries, requiring configuration to use. Spawns an external process (managed by the OpenBao server) to handle plugin requests. |

Reviewer 3: ```suggestion
| **External**   | Distributed as stand-alone binaries, requiring installation and configuration to use. Spawns an external process (managed by the OpenBao server) to handle plugin requests. |
```

> | **Builtin**    | Distributed as part of the main `bao` binary artifact. No external process is started.                                                                     |

Reviewer: I think builtin -> external instead of external -> builtin is the more natural order.

For example, [`secrets-pki`](/docs/secrets/pki) is a **builtin** **secret**
plugin, but [`secrets-aws`](https://github.com/openbao/openbao-plugins/releases?q=secrets-aws-&expanded=true)
is an **external** **secret** plugin. [`auth-aws`](https://github.com/openbao/openbao-plugins/releases?q=auth-aws-&expanded=true)
also exists and is an **external** **auth** plugin.

## Usage

OpenBao since [v2.5.x](/community/release-notes/2-5-0/#v250) supported
[declarative plugin configuration](/docs/configuration/plugins/). Three
mandatory configuration [parameters](/docs/configuration/#parameters) are
required for these demos:

```hcl
// Only plugins from the specified directory will be loaded. This helps
// to prevent arbitrary binaries from being executed and leading to
// generalized RCE.
plugin_directory = "/openbao/plugins"

// Setting this allows the server process to automatically download new
// plugins if they're not registered. Otherwise OpenBao will not contact
// the internet by default. This will occur on all nodes (active and
// standby).
plugin_auto_download = true

// Setting this allows automatic registration of the plugin into the

> // registry. This is allows operators to immediately use the plugin

Reviewer 3: ```suggestion
// catalog. This is allows operators to immediately use the plugin
```

// without having to manually register it using API commands:
//
// https://openbao.org/api-docs/system/plugins-catalog/
plugin_auto_register = true
```

With that in mind, adding an external plugin is greatly simplified:

```hcl
// Defines a new secret engine. This follows the convention established
// by HashiCorp of <project>-plugin-<type>-<name>.
plugin "secret" "aws" {
    // ## Distribution Information ##

    // Image is the registry's address. It is strongly recommended to pin to
    // specific SHA256 hashes rather than using a tag which may be moved. In
    // production environments, pull from a local registry mirror instead of
    // a public registry.
    //
    // See also: https://github.com/openbao/openbao-plugins/pkgs/container/openbao-plugin-secrets-aws/1056268559?tag=v0.3.1
    image = "ghcr.io/openbao/openbao-plugin-secrets-aws"

>     // The version of the plugin itself; must be a semantic version and also a
>     // tag on the above registry.

Reviewer 3: 1. Does OpenBao enforce that this is a "semantic version"? What if my plugin has version `2026.07.1`?
2. (General question) Do plugin binaries somehow report their version to OpenBao? Or does OpenBao take the plugin version from what the operator registers in the catalog? Aka, could the operator register a plugin as version 2.0 even though the binary is compiled from 1.0?
 
```suggestion
    // The version of the plugin itself; must be a tag on the above registry.
```

Reviewer: I think 

<img width="236" height="137" alt="image" src="https://github.com/user-attachments/assets/1c761001-0f72-476f-a580-9b3fccaa1e78" />

may be the adequate answer to these questions, but I'd need to doublecheck.

Reviewer: ```
self-initialization failed: failed to set up declarative plugins: [plugin 0 (name: aws / type: secret)]: failed to register new plugin: plugin version mismatch: aws reported version (v0.2.0) did not match requested version (v0.2.0@sha256:74c77f613f258f155054c9219ea918d5fd353ae413430cbf1587d5506afee0e4)
```

Author: Yeah, Vault was fond of forcing semantic versions in a number of places, since they write a Go semantic version library. 😆

Author: > Does OpenBao enforce that this is a "semantic version"? What if my plugin has version `2026.07.1`?

`2026.07.1` I believe is a valid semantic version. What you couldn't do is `20260701`.

    //
    // See also: https://semver.org/

>     version = "v0.3.1"

Reviewer 3: > It is strongly recommended to pin to specific SHA256 hashes rather than using a tag which may be moved.

1. Can I really pin the hash in the image field? I assumed that OpenBao merges `image` and `version`. Can you give an example of how to pin the SHA-256 digest in the `image` field?
2. Is pinning the image itself necessary? The `sha256sum` field already pins the binary inside the image, isn't that sufficient?

Reviewer: I concur with @thgoebel:

1. I'm not sure it's even possible either, at least not within the `image` field.
2. Yeah, you should just be pinning `sha256sum` and use a semver tag only for the OCI artifact. In fact, I'd mention that it is _not_ necessary (or even possible?) to pin the artifact itself separately for this reason.

Author: Yeah I wrote this first and then tested it only to realize that we can't pin sha256 because the image tag is the version. I definitely think we should either support an explicit image_tag field that would let us take a OCI image hash or just place it into the existing `image` parameter.

Author: Ah, I realize my issue now, my testing config snippets under `/tmp` weren't synced into the post itself, I had edited the wrong ones. :facepalm:

[DEBUG] plugins: local platform: plugin=secret-aws platform=linux/amd64
[DEBUG] plugins: extracting plugin from OCI image: plugin=secret-aws target=/openbao/plugins/.oci-cache/secret-aws/641ae185/openbao-plugin-secrets-aws binary=openbao-plugin-secrets-aws
[INFO]  plugins: found plugin binary in OCI image: plugin=secret-aws entry=openbao-plugin-secrets-aws size=25161890
[DEBUG] plugins: successfully extracted plugin binary: plugin=secret-aws path=/openbao/plugins/.oci-cache/secret-aws/641ae185/openbao-plugin-secrets-aws size=25161890
[INFO]  plugins: successfully downloaded and validated plugin: plugin=secret-aws cached_path=/openbao/plugins/.oci-cache/secret-aws/641ae185/openbao-plugin-secrets-aws symlink_path=/openbao/plugins/secret-aws-v0.3.1 hash=641ae1858c0f660b4c3761286d627e1dcfd33dc35e7c83eb81ca9050b5bdc6d8
[INFO]  plugins: OCI plugin downloading completed
```

It will now show up in the registry:

```shell
$ bao read -field=secret sys/plugins/catalog
[aws kubernetes kv ldap openldap pki rabbitmq ssh totp transit]
```

and the plugin can be enabled:

```shell
$ bao secrets enable aws

> Success! Enabled the aws secrets engine at: aws/

Reviewer 2: maybe let's mention `path`? otherwise this reads as: the name of the plugin is always the path of the mount for that plugin

```

Looking at the storage layout, we see that the plugin was downloaded and
placed in the plugin folder as requested:

```shell
$ tree -alh plugins/
[  100]  plugins/
├── [   66]  .oci-cache
│   └── [   16]  secret-aws
│       └── [   52]  641ae185
│           └── [  24M]  aws
└── [   80]  secret-aws-v0.3.1 -> .oci-cache/secret-aws/641ae185/aws
```

:::info

Even if you don't want to automatically download plugins, you can use the CLI

> helper and still benefit from declarative configuration:

Reviewer: ```suggestion
If you don't want to automatically download plugins at server startup, you can use the separate `bao plugin init` command and still benefit from declarative configuration:
```

Author: @satoqz Doesn't this also occur at SIGHUP time though?

Reviewer: @cipherboy sure, what sounded wrong to me was the implication that `bao plugin init` isn't "automatically downloading plugins". I'd say it still falls into the bucket of automatic downloads; what matters more is expressing "if you don't want `bao server` to somehow do it as part of its lifecycle".

Author: @satoqz Very well, I've revised again and pushed.

```
$ bao plugin init -config /openbao
[INFO]  Plugin directory: /openbao/plugins
[INFO]  Found 1 OCI plugin(s) in configuration
[INFO]  downloading plugin from OCI registry: plugin=secret-aws url=ghcr.io/openbao/openbao-plugin-secrets-aws:v0.3.1
[INFO]  found plugin binary in OCI image: plugin=secret-aws entry=openbao-plugin-secrets-aws size=25161890
[INFO]  successfully downloaded and validated plugin: plugin=secret-aws cached_path=/openbao/plugins/.oci-cache/secret-aws/641ae185/openbao-plugin-secrets-aws symlink_path=/openbao/plugins/secret-aws-v0.3.1 hash=641ae1858c0f660b4c3761286d627e1dcfd33dc35e7c83eb81ca9050b5bdc6d8
```

:::

This pattern applies for auth and database plugins. When using auto-unseal
plugins, note that the plugin will not show up in the registry but will

> otherwise be usable at startup:

Reviewer: ```suggestion
This pattern applies for secrets, auth and database plugins. When using KMS
plugins, note that the plugin will not show up in the registry but will
otherwise be usable at startup:
```

Reviewer: Would also be good to mention what "showing up in the registry" means. Probably more like "show up under plugin catalog endpoints".

Author: I added `(shown above)` to refer the reader back up.

contribute to the ecosystem!

::::

<!-- truncate -->

## Overview

Nearly everything within OpenBao is pluggable. OpenBao supports several types of plugins currently:

| Type            | Description                                                                                                                                                                   |
| :-------------- | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Auth**        | Allow bringing sources of identity--JWTs, certificates, passwords--and exchanging them for OpenBao tokens.                                                                    |
| **Secret**      | Allow generating credentials (whether passwords for databases, PKI certificates, static secrets, and more) and protected through OpenBao's tokens and authorization policies. |
| **Database**    | Implement database-specific logic for updating and managing access for use with the `database` dynamic secret engine.                                                         |

> | **KMS**         | Implement a specialized interfaces for using HSM and KMS devices (such as YubiHSM or GCP CloudKMS) for auto-unseal (and external keys in v2.6.0).                             |

Reviewer: ```suggestion
| **KMS**         | Implement specialized interfaces for using HSM and KMS devices (such as YubiHSM or GCP CloudKMS) for auto-unseal (and external keys in the upcoming v2.7.0).                  |
```

| **KMS**         | Implement a generic interface for using HSm and KMS devices within OpenBao's secret engines. _Coming in OpenBao v2.7.x_.                                                      |

Plugins have two primary classifications:

| Classification | Description                                                                                                                                                |
| :------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **External**   | Distributed as stand-alone binaries, requiring configuration to use. Spawns an external process (managed by the OpenBao server) to handle plugin requests. |
| **Builtin**    | Distributed as part of the main `bao` binary artifact. No external process is started.                                                                     |

For example, [`pki`](/docs/secrets/pki) is a **builtin** **secret** plugin, but

> [`secrets-aws`](https://github.com/openbao/openbao-plugins/releases?q=secrets-aws-&expanded=true)

Reviewer: ```suggestion
For example, [`secrets-pki`](/docs/secrets/pki) is a **builtin** **secret** plugin, but
[`secrets-aws`](https://github.com/openbao/openbao-plugins/releases?q=secrets-aws-&expanded=true)
```

Should it be this for consistency?

Author: But it is builtin and we don't yet build them as external... We could refer to `aws`, but then they'd have to resolve the ambiguity between the auth and the secrets engines. :thinking: 

I guess Kubernetes is still builtin and is ambiguous.

2026-07-29T14:24:19.255-0500 [INFO]  found plugin binary in OCI image: plugin=secret-openbao-plugin-secrets-aws entry=openbao-plugin-secrets-aws size=25161890
2026-07-29T14:24:19.520-0500 [INFO]  successfully downloaded and validated plugin: plugin=secret-openbao-plugin-secrets-aws cached_path=/openbao/plugins/.oci-cache/secret-openbao-plugin-secrets-aws/641ae185/openbao-plugin-secrets-aws symlink_path=/openbao/plugins/secret-openbao-plugin-secrets-aws-v0.3.1 hash=641ae1858c0f660b4c3761286d627e1dcfd33dc35e7c83eb81ca9050b5bdc6d8
```

:::

This pattern applies for auth and database plugins. When using auto-unseal
plugins, note that the plugin will not show up in the registry but will
otherwise be usable at startup:

```hcl
plugin "kms" "openbao-plugin-kms-pkcs11" {
    image = "ghcr.io/openbao/openbao-plugin-kms-pkcs11"
    version = "v0.1.0"
    binary_name = "openbao-plugin-kms-pkcs11"
    sha256sum = "55245882727535579e710672f0eae1bcdddc846006db857baaa6e09e33d40faf"
}

seal "openbao-plugin-kms-pkcs11" {
    lib = "/usr/lib64/softhsm/libsofthsm.so"
    token_label = "OpenBao"
    pin = "4321"
    key_label = "bao-root-key-rsa"
    rsa_oaep_hash = "sha1"

> }

Reviewer: ```suggestion
plugin "kms" "pkcs11" {
    image = "ghcr.io/openbao/openbao-plugin-kms-pkcs11"
    version = "v0.1.0"
    binary_name = "openbao-plugin-kms-pkcs11"
    sha256sum = "55245882727535579e710672f0eae1bcdddc846006db857baaa6e09e33d40faf"
}

seal "pkcs11" {
    lib = "/usr/lib64/softhsm/libsofthsm.so"
    token_label = "OpenBao"
    pin = "4321"
    key_label = "bao-root-key-rsa"
    rsa_oaep_hash = "sha1"
}
```

Is the far cleaner style IMO. Also applies to other plugin types.

Author: @satoqz Hmm, I do think we should keep the `type` prefix as there's a lot of duplicates (cloud providers for instance).

Reviewer: @cipherboy that's why you specify the plugin type right before it, no?

Author: @satoqz Ok, fair :laughing:

```

## Updating

When updating a plugin, specify the block multiple times with different
versions:

```hcl
plugin "secret" "openbao-plugin-secrets-aws" {
    image = "ghcr.io/openbao/openbao-plugin-secrets-aws"
    version = "v0.3.1"
    binary_name = "openbao-plugin-secrets-aws"
    sha256sum = "641ae1858c0f660b4c3761286d627e1dcfd33dc35e7c83eb81ca9050b5bdc6d8"
}

plugin "secret" "openbao-plugin-secrets-aws" {
    image = "ghcr.io/openbao/openbao-plugin-secrets-aws"
    version = "<new update version>"
    binary_name = "openbao-plugin-secrets-aws"
    sha256sum = "<new update sha>"
}
```

This will allow OpenBao to register two versions of the plugin. Mounts can be
individually updated to the latest plugin version and operators can have

> granular control over when that occurs.

Reviewer: This might be a bit too shallow. You'd want to mention when to make that config update, how to make openbao load it, how to then update mounts and reload... Perhaps it's better to just update the plugin upgrade guide to mention declarative plugins and link to that.

Author: @satoqz Fair, I'll wait for the other PR to merge and then link to that and publish this.
